src/usr.bin/tip/remote.5
.\"	$ssdlinux: remote.5,v 1.1 2004/11/12 15:22:13 kanoh Exp $
.\"	$NetBSD: remote.5,v 1.11 2003/08/07 10:31:17 agc Exp $
.\"
.\" Copyright (c) 1983, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)remote.5	8.1 (Berkeley) 6/5/93
.\"
.Dd June 5, 1993
.Dt REMOTE 5
.Os
.Sh NAME
.Nm remote
.Nd remote host description file
.Sh DESCRIPTION
The systems known by
.Xr tip 1
and their attributes are stored in an
.Tn ASCII
file which
is structured somewhat like the
.Xr termcap 5
file.
Each line in the file provides a description for a single
.Em system .
Fields are separated by a colon
.Pq Dq \&: .
Lines ending in a \e character with an immediately following newline are
continued on the next line.
.Pp
The first entry is the name(s) of the host system.
If there is more than one name for a system, the names are separated
by vertical bars.
After the name of the system comes the fields of the description.
A field name followed by an `=' sign indicates a string value follows.
A field name followed by a `#' sign indicates a following numeric value.
.Pp
Entries named
.Dq tip*
and
.Dq cu*
are used as default entries by
.Xr tip 1 ,
and the
.Xr cu 1
interface to
.Xr tip 1 ,
as follows.
When
.Xr tip 1
is invoked with only a phone number, it looks for an entry
of the form
.Dq tip300 ,
where 300 is the baud rate with
which the connection is to be made.
When the
.Xr cu 1
interface is used, entries of the form
.Dq cu300
are used.
.Sh CAPABILITIES
Capabilities are either strings (str), numbers (num), or boolean flags (bool).
A string capability is specified by
.Em "capability=value" ;
for example,
.Dq Li dv=/dev/harris .
A numeric capability is specified by
.Em "capability#value" ;
for example,
.Dq Li xa#99 .
A boolean capability is specified by simply listing the capability.
.Bl -tag -width indent
.It Cm \&at
(str)
Auto call unit type.
.It Cm \&br
(num)
The baud rate used in establishing
a connection to the remote host.
This is a decimal number.
The default baud rate is 300 baud.
.It Cm \&cm
(str)
An initial connection message to be sent to the remote host.
For example, if a host is reached through a port selector, this might
be set to the appropriate sequence required to switch to the host.
.It Cm \&cu
(str)
Call unit if making a phone call.
Default is the same as the `dv' field.
.It Cm \&dc
(bool)
This host is directly connected, and tip should not expect carrier detect
to be high, nor should it exit if carrier detect drops.
.It Cm \&di
(str)
Disconnect message sent to the host when a
disconnect is requested by the user.
.It Cm \&du
(bool)
This host is on a dial-up line.
.It Cm \&dv
(str)
.Ux
device(s) to open to establish a connection.
If this file refers to a terminal line,
.Xr tip 1
attempts to perform an exclusive open on the device to ensure only
one user at a time has access to the port.
.It Cm \&el
(str)
Characters marking an end-of-line.
The default is
.Dv NULL .
`~' escapes are only
recognized by
.Xr tip 1
after one of the characters in `el',
or after a carriage-return.
.It Cm \&fs
(str)
Frame size for transfers.
The default frame size is equal to
.Dv BUFSIZ .
.It Cm \&hd
(bool)
The host uses half-duplex communication, local
echo should be performed.
.It Cm \&hf
(bool)
Use hardware (RTS/CTS) flow control.
.It Cm \&ie
(str)
Input end-of-file marks.
The default is
.Dv NULL .
.It Cm \&oe
(str)
Output end-of-file string.
The default is
.Dv NULL .
When
.Xr tip 1
is transferring a file, this
string is sent at end-of-file.
.It Cm \&pa
(str)
The type of parity to use when sending data
to the host.
This may be one of
.Sy even ,
.Sy odd ,
.Sy none ,
.Sy zero
(always set bit 8 to zero),
.Sy one
(always set bit 8 to one).
The default is even parity.
.It Cm \&pn
(str)
Telephone number(s) for this host.
If the telephone number field contains an @ sign,
.Xr tip 1
searches the file
.Pa /etc/phones
file for a list of telephone numbers;
see
.Xr phones 5 .
.It Cm \&tc
(str)
Indicates that the list of capabilities is continued in the named
description.
This is used primarily to share common capability information.
.El
.Pp
Here is a short example showing the use of the capability continuation
feature:
.Bd -literal
UNIX-1200:\e
:dv=/dev/cau0:el=^D^U^C^S^Q^O@:du:at=ventel:ie=#$%:oe=^D:br#1200:
arpavax|ax:\e
:pn=7654321%:tc=UNIX-1200
.Ed
.Sh FILES
.Bl -tag -width /etc/remote -compact
.It Pa /etc/remote
The
.Nm remote
host description file
resides in
.Pa /etc .
.El
.Sh SEE ALSO
.Xr tip 1 ,
.Xr phones 5
.Sh HISTORY
The
.Nm
file format appeared in
.Bx 4.2 .


src/ChangeLog
$ssdlinux: ChangeLog,v 1.492 2005/03/20 09:18:33 kanoh Exp $

Sun Mar 20 18:15:57 JST 2005  Tamotsu Kanoh <kanoh@kanoh.org>
	* Update GCC to version 3.3.5.

Fri Mar 18 11:27:00 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update findutils to version 4.2.20.

Fri Mar 18 09:57:30 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* 0.3-20050318 Release.

Wed Mar 16 17:00:49 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update strace to version 4.5.10.

Wed Mar 16 15:48:34 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update postfix to version 2.2.1.

Tue Mar 15 11:32:34 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update gettext to version 0.14.3.

Tue Mar 15 10:18:52 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update psmisc to version 21.6.

Mon Mar 14 12:48:35 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Updated bind to version 9.2.5.

Mon Mar 14 10:15:00 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update ncurses to version 5.4-20050312.

Fri Mar 11 16:43:48 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update tnftp to version 20050103.

Fri Mar 11 09:58:54 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* 0.3-20050311 Release.

Thu Mar 10 17:44:49 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update postfix to version 2.2.0.

Thu Mar 10 16:11:46 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update iptables to version 1.3.1.

Thu Mar 10 14:14:08 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update openssh to version 4.0p1.

Mon Mar  7 10:15:38 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update ncurses to version 5.4-20050305.

Fri Mar  4 18:12:35 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* 0.3-20050304 Release.

Wed Mar  2 15:04:15 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update e2fsprogs to version 1.36.

Wed Mar  2 10:45:01 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update dhcp to version 3.0.2.

Mon Feb 28 10:14:46 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update ncurses to version 5.4-20050226.

Fri Feb 25 16:23:38 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update ex to version 050225.

Fri Feb 25 16:05:29 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update gettext to version 0.14.2.

Fri Feb 25 09:46:36 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* 0.3-20050225 Release.

Mon Feb 21 11:58:44 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update zlib to version 1.2.2.

Mon Feb 21 11:28:09 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update findutils to version 4.2.18.

Mon Feb 21 10:57:22 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update ncurses to version 5.4-20050219.

Fri Feb 18 10:39:24 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* 0.3-20050218 Release.

Tue Feb 15 18:33:14 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* 0.3-20050215 Release.

Mon Feb 14 12:03:37 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update iptables to version 1.3.0.

Mon Feb 14 11:03:45 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update file to version 4.13.

Mon Feb 14 10:40:03 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update libtool to version 1.5.14.

Mon Feb 14 10:33:00 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update automake to version 1.9.5.

Mon Feb 14 10:28:42 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update ncurses to version 5.4-20050212.

Thu Feb 10 18:02:01 JST 2005  Jun Todoroki <todoroki@plathome.co.jp>
	* Update usbustils to version 0.70 & add libusb-0.1.9.

Wed Feb  9 11:34:42 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update grub to version 0.96.

Wed Feb  9 14:36:01 JST 2005  Jun Todoroki <todoroki@plathome.co.jp>
	* Update madwifi to version 20050209.

Mon Feb  7 14:20:34 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update strace to version 4.5.9.

Mon Feb  7 13:47:22 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update sudo to version 1.6.8p7.

Mon Feb  7 11:17:34 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update cvs to version 1.11.19.

Mon Feb  7 10:47:05 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update libtool to version 1.5.12.

Mon Feb  7 10:38:15 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update ncurses to version 5.4-20050205.

Mon Jan 31 12:29:41 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update sed to version 4.1.4.

Mon Jan 31 12:19:56 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update ncurses to version 5.4-20050129.

Mon Jan 31 11:29:54 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update findutils to version 4.2.15.

Sat Jan 29 18:05:07 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* 0.3-20050129 Release.

Thu Jan 27 15:57:02 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update util-linux to version 2.12q.

Thu Jan 27 12:41:46 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update shadow to version 4.0.7.

Thu Jan 27 11:38:48 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update procps to version 3.2.5.
	
Thu Jan 27 11:18:56 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update ex to version 050119.

Thu Jan 27 11:01:33 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update dump to version 0.4b39.

Thu Jan 27 10:36:16 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update sed to version 4.1.3.

Wed Jan 26 17:08:50 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update ncurses to version 5.4-20050122.

Fri Jan 21 23:18:22 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* 0.3-20050121 Release.

Wed Jan 19 11:18:33 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
	* Update a patch for sendmail (Errata: 2005-01-17).

Mon Jan 17 11:04:35 JST 2005  Atsushi Yamagata <yamagata@plathome.co.jp>
        * Apply a patch for sendmail (Errata: 2005-01-15).


src/usr.bin/tip/remote.c
/*	$ssdlinux: remote.c,v 1.1 2004/11/12 15:22:13 kanoh Exp $	*/
/*	$NetBSD: remote.c,v 1.11 2004/04/23 22:24:34 christos Exp $	*/

/*
 * Copyright (c) 1992, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#if 0
#ifndef lint
__COPYRIGHT("@(#) Copyright (c) 1992, 1993\n\
	The Regents of the University of California.  All rights reserved.\n");
#endif /* not lint */

#ifndef lint
#if 0
static char sccsid[] = "@(#)remote.c	8.1 (Berkeley) 6/6/93";
#endif
__RCSID("$NetBSD: remote.c,v 1.11 2004/04/23 22:24:34 christos Exp $");
#endif /* not lint */
#endif

#include "pathnames.h"
#include "tip.h"
#include "getcap.h"

/*
 * Attributes to be gleened from remote host description
 *   data base.
 */
static char **caps[] = {
	&AT, &DV, &CM, &CU, &EL, &IE, &OE, &PN, &PR, &DI,
	&ES, &EX, &FO, &RC, &RE, &PA
};

static const char *capstrings[] = {
	"at", "dv", "cm", "cu", "el", "ie", "oe", "pn", "pr",
	"di", "es", "ex", "fo", "rc", "re", "pa", 0
};

static const char	*db_array[3] = { _PATH_REMOTE, 0, 0 };

#define cgetflag(f)	(cgetcap(bp, f, ':') != NULL)

static	void	getremcap __P((char *));

static char tiprecord[] = "tip.record";
static char wspace[] = "\t\n\b\f";

static void
getremcap(host)
	char *host;
{
	const char **p;
	char ***q;
	char *bp;
	char *rempath;
	int   status;

	rempath = getenv("REMOTE");
	if (rempath != NULL) {
		if (*rempath != '/')
			/* we have an entry */
			cgetset(rempath);
		else {	/* we have a path */
			db_array[1] = rempath;
			db_array[2] = _PATH_REMOTE;
		}
	}
	if ((status = cgetent(&bp, db_array, host)) < 0) {
		if (DV ||
		    (host[0] == '/' && access(DV = host, R_OK | W_OK) == 0)) {
			CU = DV;
			HO = host;
			HW = 1;
			DU = 0;
			if (!BR)
				BR = DEFBR;
			FS = DEFFS;
			return;
		}
		switch(status) {
		case -1:
			fprintf(stderr, "tip: unknown host %s\n", host);
			break;
		case -2:
			fprintf(stderr, 
			    "tip: can't open host description file\n");
			break;
		case -3:
			fprintf(stderr, 
			    "tip: possible reference loop in host description file\n");
			break;
		}
		exit(3);
	}

	for (p = capstrings, q = caps; *p != NULL; p++, q++)
		if (**q == NULL)
			cgetstr(bp, *p, *q);
	if (!BR && (cgetnum(bp, "br", &BR) == -1))
		BR = DEFBR;
	if (cgetnum(bp, "fs", &FS) == -1)
		FS = DEFFS;
	if (DU < 0)
		DU = 0;
	else
		DU = cgetflag("du");
	if (DV == NULL) {
		fprintf(stderr, "%s: missing device spec\n", host);
		exit(3);
	}
	if (DU && CU == NULL)
		CU = DV;
	if (DU && PN == NULL) {
		fprintf(stderr, "%s: missing phone number\n", host);
		exit(3);
	}

	HD = cgetflag("hd");

	/*
	 * This effectively eliminates the "hw" attribute
	 *   from the description file
	 */
	if (!HW)
		HW = (CU == NULL) || (DU && equal(DV, CU));
	HO = host;
	/*
	 * see if uppercase mode should be turned on initially
	 */
	if (cgetflag("ra"))
		setboolean(value(RAISE), 1);
	if (cgetflag("ec"))
		setboolean(value(ECHOCHECK), 1);
	if (cgetflag("be"))
		setboolean(value(BEAUTIFY), 1);
	if (cgetflag("nb"))
		setboolean(value(BEAUTIFY), 0);
	if (cgetflag("sc"))
		setboolean(value(SCRIPT), 1);
	if (cgetflag("tb"))
		setboolean(value(TABEXPAND), 1);
	if (cgetflag("vb"))
		setboolean(value(VERBOSE), 1);
	if (cgetflag("nv"))
		setboolean(value(VERBOSE), 0);
	if (cgetflag("ta"))
		setboolean(value(TAND), 1);
	if (cgetflag("nt"))
		setboolean(value(TAND), 0);
	if (cgetflag("rw"))
		setboolean(value(RAWFTP), 1);
	if (cgetflag("hd"))
		setboolean(value(HALFDUPLEX), 1);
	if (cgetflag("dc"))
		DC = 1;
	if (RE == NULL)
		RE = tiprecord;
	if (EX == NULL)
		EX = wspace;
	if (ES != NULL)
		vstring("es", ES);
	if (FO != NULL)
		vstring("fo", FO);
	if (PR != NULL)
		vstring("pr", PR);
	if (RC != NULL)
		vstring("rc", RC);
	if (cgetnum(bp, "dl", &DL) == -1)
		DL = 0;
	if (cgetnum(bp, "cl", &CL) == -1)
		CL = 0;
	if (cgetnum(bp, "et", &ET) == -1)
		ET = 10;
}

char *
getremote(host)
	char *host;
{
	char *cp;
	static char *next;
	static int lookedup = 0;

	if (!lookedup) {
		if (host == NULL && (host = getenv("HOST")) == NULL) {
			fprintf(stderr, "tip: no host specified\n");
			exit(3);
		}
		getremcap(host);
		next = DV;
		lookedup++;
	}
	/*
	 * We return a new device each time we're called (to allow
	 *   a rotary action to be simulated)
	 */
	if (next == NULL)
		return (NULL);
	if ((cp = strchr(next, ',')) == NULL) {
		DV = next;
		next = NULL;
	} else {
		*cp++ = '\0';
		DV = next;
		next = cp;
	}
	return (DV);
}
